libherokubuildpack/src/output/background.rs

```diff
@@ -0,0 +1,213 @@
+/// forever.
+#[must_use]
+pub(crate) fn print_interval<W>(
+    mut buffer: W,
+    interval: Duration,
+    start: String,
+    tick: String,
+    end: String,
+) -> state::PrintGuard<W>
+where
+    W: Write + Send + Debug + 'static,
+{
+    let (sender, receiver) = channel::<()>();
+
+    let join_handle = std::thread::spawn(move || {
+        write!(buffer, "{start}").expect("Internal error");
+        buffer.flush().expect("Internal error");
+
+        loop {
+            write!(buffer, "{tick}").expect("Internal error");
+            buffer.flush().expect("Internal error");
+
+            if receiver.recv_timeout(interval).is_ok() {
+                break;
+            }
+        }
+
+        write!(buffer, "{end}").expect("Internal error");
+        buffer.flush().expect("Internal error");
+
+        buffer
+    });
+
+    state::PrintGuard::new(join_handle, sender)
+}
+
+pub(crate) mod state {
+    use std::fmt::Debug;
+    use std::io::Write;
+    use std::sync::mpsc::Sender;
+    use std::thread::JoinHandle;
+
+    /// Holds the reference to the background printer
+    ///
+    /// Ensures that the dot printer is stopped in the event of an error. By signaling
+    /// it and joining when this struct is dropped.
+    ///
+    /// Gives access to the original io/buffer struct passed to the background writer
+    /// when the thread is manually stopped.
+    ///
+    /// # Panics
+    ///
+    /// Updates to this code need to take care to not introduce a panic. See
+    /// documentation in `PrintGuard::stop` below for more details.
+    #[derive(Debug)]
+    pub(crate) struct PrintGuard<W>
+    where
+        W: Write + Debug,
+    {
+        /// Holds the handle to the thread printing dots
+        ///
+        /// Structs that implement `Drop` must ensure a valid internal state at
+        /// all times due to E0509. The handle is wrapped in an option to allow the
+        /// inner value to be removed while preserving internal state.
+        join_handle: Option<JoinHandle<W>>,
+
+        /// Holds the signaling method to tell the background printer
+        /// to stop emitting.
+        stop_signal: Sender<()>,
+    }
+
+    impl<W> Drop for PrintGuard<W>
+    where
+        W: Write + Debug,
+    {
+        fn drop(&mut self) {
+            // A note on correctness. It might seem that it's enough to signal the thread to
+            // stop, that we don't also have to join and wait for it to finish, but that's not
+            // the case. The printer can still emit a value after it's told to stop.
+            //
+            // When that happens the output can appear in the middle of another output, such
+            // as an error message if a global writer is being used such as stdout.
+            // As a result we have to signal AND ensure the thread is stoped before
+            // continuing.
+            if let Some(join_handle) = self.join_handle.take() {
```

Comment:
Random passer-by with a random sanity check; is it desirable to do nothing if the `Option` is `None` on drop, or is that an invalid state that should fail noisily?

Reply:
Thanks for the comments and cursory look!

It should be `None` if the `fn stop` is called as that's a consuming interface https://github.com/heroku/libcnb.rs/pull/753/files/87f4d6ca2ada895445c2899045eccd8468befcf0#diff-ac625dd49f4a759f34b78c2641dfe4e1b4b9b332b1f77c37caa18138a99c5b3eR145.

Possibly I can do a better job cross-referencing that functionality in the documentation as someone else suggested the same thing in the lobste.rs thread.